Document generate HTTP endpoint (#6412)
triton-inference-server/server: docs/protocol/extension_generate.md

#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Generate Extension

This document describes Triton's generate extension. The generate

> extension provides a simple text oriented endpoint schema for interacting

Reviewer: Nit: text-oriented

> with LLM models. This generate endpoint is specific to HTTP/REST frontend.

Reviewer 2: ```suggestion
with large language models (LLMs). This generate endpoint is specific to HTTP/REST frontend.
```

## HTTP/REST

> In all JSON schemas shown in this document `$number`, `$string`, `$boolean`,

Reviewer: Nit: "All JSON schemas..." (additional word: in)

Author: Maybe below? The subject of the sentence is the types.
```suggestion
In all JSON schemas shown in this document, `$number`, `$string`, `$boolean`,
```

`$object` and `$array` refer to the fundamental JSON types. #optional
indicates an optional JSON field.

Triton exposes the generate endpoint at the following URLs. The client may use
HTTP POST request to different URLs for different response behavior, the
endpoint will return the generate results on success or an error in the case of
failure.

```
POST v2/models/${MODEL_NAME}[/versions/${MODEL_VERSION}]/generate

POST v2/models/${MODEL_NAME}[/versions/${MODEL_VERSION}]/generate_stream
```

### generate v.s. generate_stream

Both URLs expect the same request JSON object, and generate the same response

> JSON object. However, `generate` return exactly 1 response JSON object, while

Reviewer 2: ```suggestion
JSON object. However, `generate` returns exactly 1 response JSON object, while
```

`generate_stream` may return various number of responses based on the inference
results. `generate_stream` returns the responses as
[Server-Sent Events](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events)
(SSE), where each response will be a "data" chunk in the HTTP response body.

> Also note that an error may be returned during inference, whereas the HTTP

Reviewer: Nit: "Also, note that an error may be returned during inference, where the HTTP response code has been set in the first response of the SSE, which can result in receiving an error object while the status code shows success (200) Therefore, the"...

response code has been set at first response of the SSE, which can result in
receiving error object while status code is success (200). Therefore the user

> must always check whether an error object is received when generating responses

Reviewer 2: Can we hyper link here to the error object?

through `generate_stream`.

### Generate Request JSON Object

The generate request object, identified as *$generate_request*, is
required in the HTTP body of the POST request. The model name and
(optionally) version must be available in the URL. If a version is not

> provided the server may choose a version based on its own policies or

Reviewer: Nit: "provided, the server..."

return an error.

    $generate_request =
    {
      "text_input" : $string,
      "parameters" : $parameters #optional

>     }

Author: Do you want this to be part of the generate schema? Seems to be de-valuing the purpose of adding reserved "parameters" field.

>     }

Reviewer: I think the parameters route might be easier for users, or at least I would find it easier as a user.

I think there could be a way to document both approaches without confusing users, if done right. This approach is good because it allows people to be more explicit with which are parameters, so the feature still has value (e.g. if an input and parameter have identical names). I think the question though should be less about what de-values a feature and more about what provides the most ease and value to users.

>     }

Reviewer 2: I'm good either way - we can add a note:

```suggestion
   > [!NOTE] > Any additional properties in the request object are passed either as parameters or tensors based on model specification. 
```

but in all examples let's use 'parameters' for consistency (i.e. not show both forms). 

I believe we'll refine this further and may either lock the schema further or make it more flexible. Since this is provisional we won't guarantee backwards compat (yet)

* "text_input" : The text input that the model should generate output from.
* "parameters" : An optional object containing zero or more parameters for this
  generate request expressed as key/value pairs. See
  [Parameters](#parameters) for more information.

> #### Parameters

Reviewer: Isn't there also the easier route that can be shown where if you have a parameter that has a distinct name from any inputs, you can simply put it as an input and it will work?

Author: See https://github.com/triton-inference-server/server/pull/6412#discussion_r1355419563, documenting this then introduces ambiguity on how to specify additional inputs. If there are two (documented) ways to achieve the same goal, when should the user pick one over the other.

The *$parameters* JSON describes zero or more “name”/”value” pairs,
where the “name” is the name of the parameter and the “value” is a
$string, $number, or $boolean.

    $parameters =
    {
      $parameter, ...
    }

    $parameter = $string : $string | $number | $boolean

> Parameters are model-specific. In other words, the accepting names and values

Reviewer: Nit: remove accepting? Is this supposed to mean something along the lines of "the keys and values are specific to the model"?

> Parameters are model-specific. In other words, the accepting names and values

Reviewer 2: ```suggestion
Parameters are model-specific. The user should check with the model specification to set the parameters.
```

varies according to the model. The user should check with the model
specification to set the parameters.

> ### Generate Response JSON Object

Reviewer: I don't know if this deviates too much from existing documentation, but would you be able to provide an easy example to save users time and help with understanding? For example:
`curl -X POST localhost:8000/v2/models/vllm_model/generate -d '{"prompt": "text", "stream": false, "temperature": 0}'`

Reviewer: An example like this one with a description could also be a good way to show how you can send both inputs and parameters quite easily.

Reviewer 3: Do we want to include "model_name" and "model_version" in output description here? I think these are returned currently.

#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Generate Extension

> **Note:** The Generate Extension is *provisional* and likely to change in future versions.

Reviewer 3: Optional, but can get a nice render for notes using this syntax:
```
> [!NOTE]
> The Generate Extension is *provisional* and likely to change in future versions.
```

ex:
![image](https://github.com/triton-inference-server/server/assets/21284872/879879b2-b59f-4d73-818a-0f1a79225767)
